doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+barriers. If one task in a parallel region (such as a ``forall`` or
+``coforall`` loop) takes a significantly longer time than the other
+tasks, the time the loop takes will be this long time, since the program
+cannot proceed with the code after the loop until all of the tasks have
+completed.
+
+How can load imbalance be identified?
+
+ * Within a parallel region, different tasks are taking very different
+   amounts of time
+ * When profiling, the program spends a lot of time waiting for tasks to
+   complete or waiting for barriers
+
+How can load imbalance be addressed?
+
+ * There are often ways to improve the algorthim to address load
+   imbalance. For example, graph partitioners are important technology
+   that can help one balance the storage of a data structure and the
+   computation that goes with it. More generally, you might be able to
+   rearrange the data and the computation to be performed, in to a
+   strategy that has better load balance.
+
+ * In some cases, :chpl:mod:`DynamicIters` might help. You can use it to
+   write parallel loops that dynamically load balance.
+
+
+Current Issues
+--------------
+
+This section contains issues that, ideally, the Chapel compiler and
+runtime would address. However, as they may come up in practice, it's
+important to be aware of them and their workarounds.
+
+Distributed Array Field Access Can Result in Unnecessary Communication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Issue: https://github.com/chapel-lang/chapel/issues/10160
+
+When a class contains a field that is a distributed array, a distributed
+``forall`` loop using a class instance will generate communication in
+order to read the field's value (repeatedly) even though the field's
+value should be a privatized distributed array.
+
+This issue can be avoided by creating a ``ref`` or ``const ref`` that
+refers to the distributed array. This ``ref`` or ``const ref`` can be
+created outside of the ``forall`` loop and reused within it to avoid the
+problem.
+
+Creating Too Many Distributed Objects / Unoptimized Slice Assignments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In principle, distributed array creation does some work on each locale. As a
+result, it's not going to go faster when adding more locales. That can
+cause performance or scaling issues if a program tries to create too many
+distributed arrays.
+
+One aspect of the current implementation is that array slices of
+distributed arrays are also distributed objects and work is required on
+each locale to create one. That means that creating a slice, as in
+``MyDistributedArray[1..10]`` can actually be quite slow.
+
+One common scenario is that data needs to be copied between regions of
+arrays. For example
+``MyDistributedArray[1..10] = MyOtherArray[11..20];``.
+The compiler can optimize this kind of
+assignment in many cases today with Array View Elision (AVE). However, in
```

Comment:
Is there a link that can go with this?

Reply:
I could link to https://chapel-lang.org/docs/usingchapel/man.html#man-array-view-elision but that hardly explains anything... so, I don't know of a useful link here.

Reply:
(Definitely not requesting that change in this PR, to be clear :D)